Linux window decorations

mikayla-maki opened the pull request:
This PR adds support for full client side decorations on X11 and Wayland

TODO:
- [x] Adjust GPUI APIs to expose CSD related information
- [x] Implement remaining CSD features (Resizing, window border, window shadow)
- [x] Integrate with existing background appearance and window transparency
- [x] Figure out how to check if the window is tiled on X11
- [x] Implement in Zed
- [x] Repeatedly maximizing and unmaximizing can panic
- [x] Resizing is strangely slow
- [x] X11 resizing and movement doesn't work for this: https://discord.com/channels/869392257814519848/1204679850208657418/1256816908519604305
- [x] The top corner can clip with current styling
- [x] Pressing titlebar buttons doesn't work
- [x] Not showing maximize / unmaximize buttons
- [x] Noisy transparency logs / surface transparency problem https://github.com/zed-industries/zed/pull/13611#issuecomment-2201685030
- [x] Strange offsets when dragging the project panel https://github.com/zed-industries/zed/pull/13611#pullrequestreview-2154606261
- [x] Shadow inset with `_GTK_FRAME_EXTENTS` doesn't respect tiling on X11 (observe by snapping an X11 window in any direction)

Release Notes:

- N/A

mikayla-maki committed:
Default platform-specific GPUI methods
Adjust hello world example for wayland
Add GPUI methods for window resizing

mikayla-maki committed:
Implement window decoration and tilings

mikayla-maki committed:
Implement a first pass at client side decorations in GPUI

mikayla-maki committed:
WIP

mikayla-maki committed:
fix

mikayla-maki committed:
WIP

mikayla-maki committed:
Implement the window border

mikayla-maki committed:
fix compiling

mikayla-maki committed:
Finish implementing the window decorations example

co-authored-by: conrad <conrad@zed.dev>

mikayla-maki committed:
clippy

mikayla-maki committed:
fix macos

mikayla-maki committed:
Implement window decorations on x11

mikayla-maki committed:
clippy

mikayla-maki committed:
WIP: handle window and renderer transparency in the presence of CSD

mikayla-maki committed:
Finish x11 window decorations for example

mikayla-maki committed:
apply background appearance changes to wayland

mikayla-maki committed:
Implement CSD for Zed

mikayla-maki committed:
Fix on other platforms

mikayla-maki committed:
Fix tests

someone13574 commented on crates/workspace/src/workspace.rs:
```diff
@@ -4165,156 +4166,161 @@ impl Render for Workspace {
+                .justify_start()
+                .items_start()
+                .text_color(colors.text)
+                .bg(colors.background)
```
```suggestion
                .bg(colors.background)
                .overflow_hidden()
```

Not sure if this would be the best place to add it, but the window content should have `.overflow_hidden()` set so that you don't get elements clipping outside the window. This doesn't fix any clipping due to corners though.

![Screenshot from 2024-06-29 23-21-38](https://github.com/zed-industries/zed/assets/81528246/56a676cb-aebd-47db-88cf-8ad4cebe6e33)

mikayla-maki committed:
Update crates/workspace/src/workspace.rs

Co-authored-by: Owen Law <81528246+someone13574@users.noreply.github.com>

apricotbucket28 commented on crates/gpui/src/platform/linux/wayland/window.rs:
```diff
@@ -311,25 +330,42 @@ impl WaylandWindowStatePtr {
                         state.fullscreen = configure.fullscreen;
                         state.maximized = configure.maximized;
+                        state.tiling = configure.tiling;
+                        configure.size = compute_outer_size(
```
I believe the `compute_outer_size` call isn't needed when maximizing?

apricotbucket28 replied:
Changing the code to the following fixes maximization/unmaximization on Wayland:
```suggestion
                        configure.size = if configure.maximized {
                            configure.size
                        } else {
                            compute_outer_size(
                                state.window_bounds,
                                state.client_area,
                                configure.size,
                            )
                        };
```

apricotbucket28 commented on crates/title_bar/src/title_bar.rs:
```diff
@@ -379,18 +380,23 @@ impl Render for TitleBar {
-                            cx.show_window_menu(ev.position)
-                        })
-                        .on_mouse_move(move |ev, cx| {
-                            if ev.dragging() {
```
Removing this check makes it impossible to press any buttons on the title bar on Wayland or to double click to maximize. It also changes to a grabbing cursor every time I click on the title bar on XWayland.

apricotbucket28 replied:
This is now fixed! :smile: The PR description could tag https://github.com/zed-industries/zed/issues/12639

apricotbucket28 reviewed (changes requested):
Great work! These are some of the issues I saw:

- Rounded window corners don't seem to be clipped properly:

	![image](https://github.com/zed-industries/zed/assets/71973804/acaffa1c-1ddf-4ec6-ae50-0c5a3725eceb)

- Only the close button is visible, but I'm on KDE which should have all buttons (haven't checked what Wayland tells the client yet)
	![image](https://github.com/zed-industries/zed/assets/71973804/27bd6893-5e0c-45a9-b8e9-01d70e6c77b7)

- Resizing on XWayland is very flickery (does the same happen on native X11?)

	https://github.com/zed-industries/zed/assets/71973804/87046733-fdf1-4d60-9c33-bfedb92c4f7d

someone13574 commented on crates/gpui/src/platform/linux/x11/window.rs:
```diff
@@ -557,6 +573,39 @@ impl X11Window {
     }
+
+    fn send_moveresize(&self, flag: u32) {
+        let state = self.0.state.borrow();
```
```suggestion
        let state = self.0.state.borrow();
        
        self.0
            .xcb_connection
            .ungrab_pointer(x11rb::CURRENT_TIME)
            .unwrap()
            .check()
            .unwrap();
```

As per the spec *"The Client MUST release all grabs prior to sending such message (except for the _NET_WM_MOVERESIZE_CANCEL message)"*. You probably should regrab the pointer after as well, but it seems to work without doing so. This fixes resizing on native x11 gnome, which otherwise fails [here](https://gitlab.gnome.org/GNOME/mutter/-/blob/main/src/compositor/meta-window-drag.c#L1860).

someone13574 commented:
> Resizing on XWayland is very flickery (does the same happen on native X11?)

Its much worst (at least on my system). Zed simply stops re-rendering when resizing and smears/resizes/distorts the content of its frames.

mikayla-maki committed:
Update crates/gpui/src/platform/linux/wayland/window.rs

Co-authored-by: apricotbucket28 <71973804+apricotbucket28@users.noreply.github.com>

mikayla-maki committed:
Update crates/gpui/src/platform/linux/x11/window.rs

Co-authored-by: Owen Law <81528246+someone13574@users.noreply.github.com>

mikayla-maki committed:
fix compile error

mikayla-maki committed:
Fix

mikayla-maki committed:
Fix non-resizing problems

mikayla-maki commented:
Resizing is still broken on both implementations, but the other issues should be fixed :)

apricotbucket28 committed:
linux: Small improvements to window decorations (#13715)

Note: This PR targets the `linux-window-decorations` branch, not `main`.

Release Notes:

- N/A

apricotbucket28 commented on crates/workspace/src/workspace.rs:
```diff
@@ -6474,3 +6482,176 @@ mod tests {
+    const ROUNDING: Pixels = px(10.0);
+    const SHADOW_SIZE: Pixels = px(10.0);
+    const BORDER_SIZE: Pixels = px(1.0);
+    let grey = rgb(0xC8C8C8);
```
```suggestion
    let color = cx.theme().colors().border;
```
I think this looks much better than hard-coding a greyish color :slightly_smiling_face:

![image](https://github.com/zed-industries/zed/assets/71973804/ca7c7f70-0aff-4b7a-820e-06655c24044e)

Also, there seem to still be some clipping issues with the two rounded window corners:
![image](https://github.com/zed-industries/zed/assets/71973804/1252be8b-f266-4d12-854d-ba271cebd84c)

someone13574 commented:
I'm getting an error message due to my Gnome X11 surface not supporting a composite alpha flag. On Wayland surfaces it does support it. This error message gets spammed whenever the window is resized from [here](https://github.com/kvark/blade/blob/21a56f780e21e4cb42c70a1dcf4b59842d1ad7f7/blade-graphics/src/vulkan/init.rs#L739).

Interestingly, this doesn't seem to cause any rendering problems because it just goes along and [alpha blends anyways](https://github.com/zed-industries/zed/blob/ce48555f8d7bf6071ce6c268fe073e4f1e1b7e9c/crates/gpui/src/platform/blade/blade_renderer.rs#L217). Also, if you disable transparency by overriding [here](https://github.com/zed-industries/zed/blob/72d0457595cd2e29cfc3be506c8a62c623865dc8/crates/gpui/src/platform/blade/blade_renderer.rs#L434), there is no change to the appearance. Does transparency really need to be enabled all the time or can it follow the old behavior of only being enabled when needed by the theme? Shadows seem to work with it disabled so I'm a bit confused why its needed.

One last thing: I won't be the only person with this problem; there is [very poor](https://vulkan.gpuinfo.org/listsurfacecompositealphamodes.php?platform=linux) coverage for this surface capability.

ConradIrwin committed:
Some stuff

Co-Authored-By: Mikayla <mikayla@zed.dev>

ConradIrwin committed:
More stuff

ConradIrwin committed:
Better cornering

apricotbucket28 commented on crates/workspace/src/workspace.rs:
```diff
@@ -6620,38 +6641,105 @@ pub fn client_side_decorations(element: impl IntoElement, cx: &mut WindowContext
                 .size_full()
                 .child(element),
         )
+        .map(|div| match decorations {
```
Was this code duplicated accidentally? :thinking:

mikayla-maki replied:
Nope, that's the bit that sets the cursor style. We had to remove it from the code above

apricotbucket28 replied:
Hmm, that makes sense. But looks like the old code above is still there:
1st: https://github.com/zed-industries/zed/blob/0395b02a881bc920e74dba916e4028653b9cb58d/crates/workspace/src/workspace.rs#L6500-L6543
2nd: https://github.com/zed-industries/zed/blob/0395b02a881bc920e74dba916e4028653b9cb58d/crates/workspace/src/workspace.rs#L6644-L6685

mikayla-maki replied:
Ooh, I see now, thank you!

apricotbucket28 commented on crates/workspace/src/workspace.rs:
```diff
@@ -6474,3 +6482,264 @@ mod tests {
+        .map(|div| match decorations {
+            Decorations::Server => div,
+            Decorations::Client { tiling, .. } => div
+                .bg(gpui::transparent_black())
```
```suggestion
```
Minor detail, but the same background is already set right above :slightly_smiling_face:

apricotbucket28 reviewed (changes requested):
Resizing on Wayland is now fixed, great! :smile: (even better than on main!)

I noticed another issue while resizing the project panel (doesn't happen while maximized)

https://github.com/zed-industries/zed/assets/71973804/60d843b9-7c7f-4067-bb77-6246775f74ae

mikayla-maki commented:
Thanks for checking it over @apricotbucket28 :D only issue we have right now is using `set_window_geometry` correctly on wayland, so that your window shadow doesn't collide with the rest of the UI :)

mikayla-maki committed:
Fix most resizing problems

mikayla-maki committed:
finish resize

mikayla-maki committed:
Merge branch 'main' into linux-window-decorations

mikayla-maki committed:
clippy / fmt

ConradIrwin committed:
Fix left pane resizing logic

ConradIrwin committed:
Only set transparent for background

ConradIrwin committed:
Revert "Only set transparent for background"

This reverts commit 23ec011afa4ad06c8611e807c55ff16c9a39de4b.

ConradIrwin committed:
Remove shadows option

ConradIrwin committed:
Fix all the problems

ConradIrwin committed:
Merge branch 'main' into linux-window-decorations

mikayla-maki committed:
Adjust titlebar UI and fix prompts

mikayla-maki committed:
Fix windows

jvoisin commented on crates/gpui/src/color.rs:
```diff
@@ -309,6 +309,16 @@ pub fn transparent_black() -> Hsla {
     }
 }
 
+/// Transparent black in [`Hsla`]
```
Transparent white I think :)

Also, can't the function be made `const fn`?

a15355447898a commented:
I'm getting very stuck resizing windows on wayland(not xwayland).
The version is[v0.143.4-pre](https://github.com/zed-industries/zed/releases/tag/v0.143.4-pre)

https://github.com/zed-industries/zed/assets/62456287/1cdadb43-d897-4acd-83ef-64ff425fb244

RustoMCSpit commented:
this may be related https://github.com/zed-industries/zed/issues/15245

RustoMCSpit commented:
double clicking the top should maximise minimise but it doesnt

RustoMCSpit commented:
> double clicking the top should maximise minimise but it doesnt

just found out it works but it's extremely janky. it only works 15 percent of the time
